File: main_project/tripod-new/src/cpp/client/TripodClient.cpp
#include <TripodData.h>
#include <TripodClient.h>
#include <ClientFactory.h>
#include <CacheClient.h>
#include <RedisCacheClient.h>
#include <CacheManagerClient.h>

namespace com {
namespace renren {
namespace tripod {

TripodClient::TripodClient(const std::string& zkAddress, const std::string& namespaceId, 
                        const std::string& businessId) : 
                        zkAddress_(zkAddress), namespaceId_(namespaceId), businessId_(businessId) {
}

CacheClientPtr TripodClient::getCacheClient() {
    if(cacheClient_ == NULL) {
        boost::upgrade_lock<boost::shared_mutex> lock(cacheClientMutex_);
        boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock1(lock);
        if(cacheClient_ == NULL) {
            cacheClient_ = ClientFactory::getInstance()->getCacheClient(zkAddress_);
        }
    }
    return cacheClient_;
}

CacheManagerClientPtr TripodClient::getCacheManagerClient() {
    if(cacheManagerClient_ == NULL) {
        boost::upgrade_lock<boost::shared_mutex> lock(cacheManagerClientMutex_);
        boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock1(lock);
        if(cacheManagerClient_ == NULL) {
            cacheManagerClient_ = ClientFactory::getInstance()->getCacheManagerClient(zkAddress_);
        }
    }
    return cacheManagerClient_;
}

DataMap TripodClient::get(const KeySeq& keys, KeySeq& missedKeys, const long timeout) {
    CacheClientPtr client = getCacheClient();
    if(client == NULL) {
        return DataMap();
    }

    DataMap res = client->get(keys, namespaceId_, businessId_, timeout);
    
    getMissedKeys(res, keys, missedKeys);

    return res;
}

DataMap TripodClient::getMissed(const KeySeq& keys, long expiredTime, bool useMaster, const long timeout) {
    CacheManagerClientPtr client = getCacheManagerClient();
    if(client == NULL) {
        return DataMap();
    }

    return client->get(keys, namespaceId_, businessId_, expiredTime, useMaster, timeout);
}

DataMap TripodClient::getWithMissed(const KeySeq& keys, long expiredTime, bool useMaster, const long timeout) {
    KeySeq missedKeys;
    DataMap res = get(keys, missedKeys, timeout);

    if(!missedKeys.empty()) {
        DataMap missedRes = getMissed(missedKeys, expiredTime, useMaster, timeout);
        res.insert(missedRes.begin(), missedRes.end());
    }
    return res;
}


bool TripodClient::load(const KeySeq& keys, long expiredTime, bool useMaster, const long timeout) {
    CacheManagerClientPtr client = getCacheManagerClient();
    if(client == NULL) {
        return false;
    }

    return client->load(keys, namespaceId_, businessId_, expiredTime, useMaster, timeout);
}

bool TripodClient::deleteCache(const KeySeq& keys, const long timeout) {
    CacheManagerClientPtr client = getCacheManagerClient();
    if(client == NULL) {
        return false;
    }
    
    return client->deleteCache(keys, namespaceId_, businessId_, timeout);
}

void TripodClient::getMissedKeys(const DataMap& res, const KeySeq& keys, KeySeq& missedKeys) {
    for(KeySeq::const_iterator it = keys.begin(); it != keys.end(); ++it) {
        if(res.find(*it) == res.end()) {
            missedKeys.push_back(*it);
        }
    }
}



RedisCacheClientPtr TripodClient::GetRedisCacheClient() {
  if(cacheClient_ == NULL) {
    boost::upgrade_lock<boost::shared_mutex> lock(redis_cache_client_mutex_);
    boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock1(lock);
    if(redis_cache_client_ == NULL) {
       MCE_DEBUG("zkAddress_" << zkAddress_ );
       redis_cache_client_ = ClientFactory::getInstance()->GetRedisCacheClient(zkAddress_);
    }
  }
  return redis_cache_client_;
}

void TripodClient::GetList(const std::string& key, StrList& value, int begin, int limit) {
  RedisCacheClientPtr client = GetRedisCacheClient();
  if (!client) {
    MCE_WARN("TripodClient::GetList() RedisCacheClient is NULL");
    return;
  }
  const StrList& list = client->GetList(key, begin, limit, namespaceId_, businessId_);
  if (!list.empty()) {
    value.insert(value.end(), list.begin(), list.end());
  }
}

void TripodClient::GetHash(const std::string& key, const StrList& fields, HashPair& value) {
  RedisCacheClientPtr client = GetRedisCacheClient();
  if (!client) {
    MCE_WARN("TripodClient::GetHash() RedisCacheClient is NULL");
    return;
  }
  int exists = client->Exists(key, namespaceId_, businessId_);
  if (exists == kReplySucess) { //命中
    bool ret = client->GetHash(key, fields, value, namespaceId_, businessId_);
  }
  else { //未命中
    return GetMissedHash(key, fields, value);
  }
}

bool TripodClient::Exists(const std::string& key) {
  RedisCacheClientPtr client = GetRedisCacheClient();
  if (!client) {
    MCE_WARN("TripodClient::GetHash() RedisCacheClient is NULL");
    return false;
  }
  int exists = client->Exists(key, namespaceId_, businessId_);
  return (exists == kReplySucess ? true : false);
}

void TripodClient::GetMissedHash(const std::string& key, const StrList& fields, HashPair& value, bool use_master) {
  CacheManagerClientPtr client = getCacheManagerClient();
  if(client == NULL) {
    MCE_WARN("TripodClient::GetMissedHash() getCacheManagerClient is NULL");
      return;
  }
  BigKeyDataMap ret_data = client->GetHashCache(key, fields, namespaceId_, businessId_, use_master);
  if (!ret_data.empty()) {
    BigKeyDataMap::const_iterator iter = ret_data.begin();
    for (; iter != ret_data.end(); ++iter) {
      value.insert(std::make_pair(iter->first, iter->second));
    }
  }
}

bool TripodClient::LoadListCache(const std::string& key, bool use_master, const long timeout) {
  CacheManagerClientPtr client = getCacheManagerClient();
  if(client == NULL) {
      return false;
  }
  return client->LoadListCache(key, namespaceId_, businessId_, use_master);
  
}

bool TripodClient::LoadHashCache(const std::string& key, bool use_master, const long timeout) {
  CacheManagerClientPtr client = getCacheManagerClient();
  if(client == NULL) {
      return false;
  }
  return client->LoadHashCache(key, namespaceId_, businessId_, use_master);
  
}

bool TripodClient::DeleteListOrHashCache(const std::string& key, const long timeout) {
  CacheManagerClientPtr client = getCacheManagerClient();
  if(client == NULL) {
      return false;
  }
  return client->DeleteListOrHashCache(key, namespaceId_, businessId_);
}

}}}

File: main_project/tripod-new/src/java/TripodClient/src/main/java/com/renren/tripod/client/TripodClient.java
package com.renren.tripod.client;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.renren.tripod.internalclient.CacheClient;
import com.renren.tripod.internalclient.CacheManagerClient;
import com.renren.tripod.internalclient.ClientFactory;
import com.renren.tripod.internalclient.RedisCacheClient;
import com.xiaonei.xce.log.Tripod;

public class TripodClient implements ITripodClient {

    private String _zkAddress = null;
    private String _namespaceId = null;
    private String _businessId = null;

    private CacheClient _cacheClient = null;
    private CacheManagerClient _cacheManagerClient = null;
    private RedisCacheClient redisCacheClient = null;

    public TripodClient(String zkAddress, String namespaceId, String businessId) {
	_zkAddress = zkAddress;
	_namespaceId = namespaceId;
	_businessId = businessId;
    }

    public Map<String, byte[]> get(List<String> keys, List<String> missedKeys) {
	CacheClient cc = getCacheClient();
	if (cc == null) {
	    Tripod.getLogger().error(
		    "TripodClient.get cacheclient == null for zkAddress "
			    + _zkAddress);
	    return null;
	}
	Map<String, byte[]> res = cc.get(keys, _namespaceId, _businessId);
	getMissedKeys(res, keys, missedKeys);
	return res;
    }

    public Map<String, byte[]> getMissed(List<String> keys, int expiredTime,
	    boolean useMaster) {
	return getMissed(keys, expiredTime, useMaster, 300);
    }

    public Map<String, byte[]> getMissed(List<String> keys, int expiredTime,
	    boolean useMaster, int timeout) {
	CacheManagerClient cc = getCacheManagerClient();
	if (cc == null) {
	    Tripod.getLogger().error(
		    "TripodClient.getMissed cachemanagerclient == null for zkAddress "
			    + _zkAddress);
	    return null;
	}
	return cc.get(keys, _namespaceId, _businessId, expiredTime, useMaster,
		timeout);
    }

    public Map<String, byte[]> getWithMissed(List<String> keys,
	    int expiredTime, boolean useMaster, int timeout) {
	List<String> missedKeys = new ArrayList<String>();
	Map<String, byte[]> res = get(keys, missedKeys);

	if (!missedKeys.isEmpty()) {
	    Map<String, byte[]> missedRes = getMissed(missedKeys, expiredTime,
		    useMaster, timeout);
	    res.putAll(missedRes);
	}
	return res;
    }

    public boolean load(List<String> keys, int expiredTime, boolean useMaster) {
	return load(keys, expiredTime, useMaster, 1000);
    }

    public boolean load(List<String> keys, int expiredTime, boolean useMaster,
	    int timeout) {
	CacheManagerClient cc = getCacheManagerClient();
	if (cc == null) {
	    Tripod.getLogger().error(
		    "TripodClient.load cachemanagerclient == null for zkAddress "
			    + _zkAddress);
	    return false;
	}
	return cc.load(keys, _namespaceId, _businessId, expiredTime, useMaster,
		timeout);
    }

    public boolean deleteCache(List<String> keys) {
	return deleteCache(keys, 1000);
    }

    public boolean deleteCache(List<String> keys, int timeout) {
	CacheManagerClient cc = getCacheManagerClient();
	if (cc == null) {
	    Tripod.getLogger().error(
		    "TripodClient.deleteCache cachemanagerclient == null for zkAddress "
			    + _zkAddress);
	    return false;
	}
	return cc.deleteCache(keys, _namespaceId, _businessId, timeout);
    }

    private List<String> getMissedKeys(Map<String, byte[]> res,
	    List<String> keys, List<String> missedKeys) {
	for (String str : keys) {
	    if (res.get(str) == null) {
		missedKeys.add(str);
	    }
	}
	return missedKeys;
    }

    private CacheClient getCacheClient() {
	if (_cacheClient == null) {
	    synchronized (this) {
		if (_cacheClient == null) {
		    _cacheClient = ClientFactory.getCachegClient(_zkAddress);
		}
	    }
	}
	return _cacheClient;
    }

    private CacheManagerClient getCacheManagerClient() {
	if (_cacheManagerClient == null) {
	    synchronized (this) {
		if (_cacheManagerClient == null) {
		    _cacheManagerClient = ClientFactory
			    .getCacheManagerClient(_zkAddress);
		}
	    }
	}
	return _cacheManagerClient;
    }

    private RedisCacheClient getRedisCacheClient() {
	if (redisCacheClient == null) {
	    synchronized (this) {
		if (redisCacheClient == null) {
		    redisCacheClient = ClientFactory
			    .getRedisCachegClient(_zkAddress);
		}
	    }
	}
	return redisCacheClient;
    }

    public Map<String, String> getHash(String key, List<String> fields) {
	Map<String, String> result = new HashMap<String, String>();
	RedisCacheClient rcc = getRedisCacheClient();
	if (rcc == null) {
	    Tripod.getLogger().error(
		    "TripodClient.getHash getRedisCacheClient() == null for zkAddress "
			    + _zkAddress);
	    System.err
		    .println("TripodClient.getHash getRedisCacheClient() == null");
	    return result;
	}
	boolean exist = rcc.exists(key, _namespaceId, _businessId);
	if (!exist) {
	    // timeout for 10 secs
	    return getMissedHash(key, fields, true, 10000);
	} else {
	    result = rcc.getHash(key, fields, _namespaceId, _businessId);
	}
	return result;
    }

    public Map<String, String> getMissedHash(String key, List<String> fields,
	    boolean useMaster, int timeout) {
	Map<String, String> result = new HashMap<String, String>();
	CacheManagerClient cmc = getCacheManagerClient();
	if (cmc == null) {
	    Tripod.getLogger().error(
		    "TripodClient.getMissedHash getCacheManagerClient() == null for zkAddress "
			    + _zkAddress);
	    System.err
		    .println("TripodClient.getMissedHash getCacheManagerClient() == null");
	}
	result = cmc.getHashCache(key, fields, _namespaceId, _businessId,
		useMaster, timeout);
	return result;
    }

    public boolean loadHash(String key, boolean useMaster, int timeout) {
	CacheManagerClient cmc = getCacheManagerClient();
	if (cmc == null) {
	    Tripod.getLogger().error(
		    "TripodClient.loadHashCache getCacheManagerClient() == null for zkAddress "
			    + _zkAddress);
	    System.err
		    .println("TripodClient.loadHashCache getCacheManagerClient() == null");
	}
	return cmc.loadHashCache(key, _namespaceId, _businessId, useMaster,
		timeout);
    }

    public boolean exists(String key) {
	RedisCacheClient rcc = getRedisCacheClient();
	if (rcc == null) {
	    Tripod.getLogger().error(
		    "TripodClient.getHash getRedisCacheClient() == null for zkAddress "
			    + _zkAddress);
	    System.err
		    .println("TripodClient.getHash getRedisCacheClient() == null");
	    return false;
	}
	boolean exist = rcc.exists(key, _namespaceId, _businessId);
	return exist;
    }
}
